src/s3/entries.rs

```diff
@@ -1,150 +1,77 @@
 use super::{to_block, to_block_raw};
 use crate::ipfs::{Block, Ipfs, KeplerParams};
 use anyhow::Result;
-use ipfs_embed::TempPin;
 use libipld::{cid::Cid, store::StoreParams, DagCbor};
+use libp2p::futures::stream::BoxStream;
 use std::{
     collections::BTreeMap,
     io::{self, Cursor, ErrorKind, Write},
-    pin::Pin,
-    task::{Context, Poll},
 };
 
-use rocket::tokio::io::{copy, AsyncRead, AsyncWrite, AsyncWriteExt, ReadBuf};
-
-pub struct IpfsReadStream {
-    store: Ipfs,
-    block: Cursor<Block>,
-    index: usize,
-    pub content: Vec<(Cid, u32)>,
+use rocket::futures::{StreamExt, TryStreamExt};
+use rocket::tokio::io::AsyncRead;
+use tokio_stream::iter;
+use tokio_util::io::{ReaderStream, StreamReader};
+
+pub type ObjectReader =
+    StreamReader<BoxStream<'static, Result<Cursor<Block>, io::Error>>, Cursor<Block>>;
+
+pub fn read_from_store(ipfs: Ipfs, content: Vec<(Cid, u32)>) -> ObjectReader {
+    let chunk_stream = Box::pin(
+        iter(content)
+            .then(move |(cid, _)| get_block(ipfs.clone(), cid))
+            .map_ok(|block| Cursor::new(block))
+            .map_err(|ipfs_err| io::Error::new(ErrorKind::Other, ipfs_err)),
+    );
+    StreamReader::new(chunk_stream)
 }
 
-impl IpfsReadStream {
-    pub fn new(store: Ipfs, content: Vec<(Cid, u32)>) -> Result<Self> {
-        let (cid0, _) = content.first().ok_or_else(|| anyhow!("Empty Content"))?;
-        let block = Cursor::new(store.get(cid0)?);
-        Ok(Self {
-            store,
-            content,
-            block,
-            index: 0,
-        })
-    }
+async fn get_block(ipfs: Ipfs, cid: Cid) -> Result<Block, ipfs::Error> {
```

Comment:
Yep, it's moving the cloned IPFS into the ObjectReader stream. We run into a whole load of lifetime issues, because I believe at the top level we need the response to be static (for Rocket?). I can have another play around with it and double check that it's impossible, but I don't think it's too big of a deal since `Ipfs` is just a lightweight facade.

Reply:
yeah should be fine, it is designed that way